src/EventStore.Core.Tests/Services/Transport/Enumerators/Enumerator.AllSubscriptionFiltered.CombinationTests.cs

```diff
@@ -576,6 +564,25 @@ public enum EventFilterType {
 				}
 			}
 
+			bool shouldCheckpoint = false;
+
+			// if we've just finished catching up
+			shouldCheckpoint |= catchingUp && SubscriptionProps.CheckpointType is
+				CheckpointType.Start or
+				CheckpointType.AtZero or
+				CheckpointType.OneBeforeLast;
+
+			// or we were live and we've received `checkpoint interval` events since the last checkpoint
+			shouldCheckpoint |= !catchingUp && numEventsSinceLastCheckpoint == CheckpointInterval;
```

Comment:
@YoEight hey, following the new semantics, the checkpoint interval is an upper bound, so `numEventsSinceLastCheckpoint must be <= CheckpointInterval`.

Say `CheckpointInterval` is 16, what the above check verifies is that if we've received 16 events since the last checkpoint, then the next thing we enumerate must be a checkpoint as we've already reached the upper bound limit.